tests/cache/test_cache.py

```diff
@@ -0,0 +1,239 @@
+        expected_hits = 0
+
+        # Generate 25 accesses to the cache, fill line after miss
+        for i in range(accesses):
+            hit = self.cache.pr_write(test_addresses[i % 3])
+            if not hit:
+                expected_misses += 1
+                # Fill the line after miss
+                self.cache.cache_line_fill(test_addresses[i % 3], MESIState.EXCLUSIVE)
+            else:
+                expected_hits += 1
+
+        self.assertEqual(self.cache.statistics.cache_reads, 0)
+        self.assertEqual(self.cache.statistics.cache_misses, expected_misses)
+        self.assertEqual(self.cache.statistics.cache_hits, expected_hits)
+        self.assertEqual(self.cache.statistics.cache_writes, accesses)
+
+    def test_basic_mesi_state(self):
+        """Test basic getting and setting of MESI state"""
+        addr = 0x1234_5678
+        
+        # Initially state should be None (line not present)
+        self.assertIsNone(self.cache.get_line_state(addr))
+        
+        # Fill line with EXCLUSIVE state
+        self.cache.cache_line_fill(addr, MESIState.EXCLUSIVE)
+        
+        # Verify we can get the state
+        self.assertEqual(self.cache.get_line_state(addr), MESIState.EXCLUSIVE)
+        
+        # Verify we can set a new state
+        self.assertTrue(self.cache.set_line_state(addr, MESIState.MODIFIED))
+        self.assertEqual(self.cache.get_line_state(addr), MESIState.MODIFIED)
+        
+        # Verify setting state of non-existent line returns False
+        non_existent_addr = 0x5555_5555
+        self.assertFalse(self.cache.set_line_state(non_existent_addr, MESIState.MODIFIED))
+    
+    def test_snoop_line(self):
+        """Test snoop line lookup functionality"""
+        addr = 0x1234_5678
+        
+        # Initially line shouldn't be present
+        self.assertIsNone(self.cache.snoop_line(addr))
+        
+        # Add line in Modified state
+        self.cache.pr_write(addr)
+        self.cache.cache_line_fill(addr, MESIState.MODIFIED)
+        
+        # Set L1 inclusion status
+        addr_fields = self.cache._Cache__decompose_address(addr)
+        cache_set = self.cache.sets[addr_fields.index]
+        way_index = cache_set.find_way_by_tag(addr_fields.tag)
+        cache_set.ways[way_index].in_l1 = True
+        
+        # Check snoop results
+        line = self.cache.snoop_line(addr)
+        self.assertIsNotNone(line)
+        self.assertEqual(line.mesi_state, MESIState.MODIFIED)
+        self.assertTrue(line.in_l1)
+        
+        # Verify snoop doesn't affect cache state
+        self.assertEqual(
+            self.cache.get_line_state(addr),
+            MESIState.MODIFIED
+        )
+        
+        # Verify modifying returned line doesn't affect cache
+        line.mesi_state = MESIState.SHARED
+        self.assertEqual(
+            self.cache.get_line_state(addr),
+            MESIState.MODIFIED
+        )
+    
+    def test_cache_line_fill(self):
```

Comment:
Do you know if it's possible to break this up into a couple of smaller tests, ideally with some shared setup? There are a number of different cases being testing in this one test. Maybe using a test suite?

We may also be at the point where we could benefit from some test setup utilities, that do things like initialize a cache, fill all ways in a set, etc. If we can consolidate the setup, that might make it easier to understand the context and purpose for each assertion.